src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -652,9 +750,76 @@ FlexCPUThread::onExecutionCompleted(shared_ptr<InflightInst> inst_ptr,
     if (inst_ptr->staticInst()->isControl()) {
         // We must have our next PC now, since this branch has just resolved
 
-        // If we are a control instruction, then we were unable to begin a
-        // fetch during decode, so we should start one now.
-        advanceInst(getNextPC());
+        if (_cpuPtr->hasBranchPredictor()) {
+            // If a branch predictor has been set, then we need to check if
+            // a prediction was made, squash instructions if incorrectly
+            // predicted, and notify the predictor accordingly.
+
+            auto it = find(inflightInsts.begin(), inflightInsts.end(),
+                           inst_ptr);
+
+            if (it == inflightInsts.end()) {
```

Comment:
`panic_if`

Reply:
I'd rather leave that as a block, because it might someday become possible, if there exist store instructions which are also branch instructions (since they can be committed and removed from the buffer before they actually have an onExecutionCompleted call)? It's an edge case but it might make it easier to deal with if that comes.